File: scripts/jtl_to_allure.py
#!/usr/bin/env python3
import csv
import json
import os
import sys
import uuid
from collections import Counter, defaultdict
from pathlib import Path

TARGET_URL = os.getenv("TARGET_URL", "https://www.blazedemo.com")
ACCEPTANCE_RPS = float(os.getenv("ACCEPTANCE_RPS", "250"))
ACCEPTANCE_P90_MS = int(float(os.getenv("ACCEPTANCE_P90_MS", "2000")))
TOP_FAILED_SAMPLES = int(float(os.getenv("ALLURE_FAILED_SAMPLES", "5")))


def safe_int(value: str, default: int = 0) -> int:
    try:
        return int(float(value))
    except Exception:
        return default


def percentile(values, pct):
    if not values:
        return 0.0
    ordered = sorted(values)
    idx = max(0, min(int(len(ordered) * pct) - 1, len(ordered) - 1))
    return float(ordered[idx])


def compute_metrics(rows):
    if not rows:
        return {
            "total": 0,
            "failed": 0,
            "error_pct": 0.0,
            "avg_ms": 0.0,
            "avg_latency_ms": 0.0,
            "p90_ms": 0.0,
            "rps": 0.0,
            "min_ts": 0,
            "max_ts": 0,
        }

    elapsed = [safe_int(r.get("elapsed", "0")) for r in rows]
    latency = [safe_int(r.get("Latency", "0")) for r in rows]
    timestamps = [safe_int(r.get("timeStamp", "0")) for r in rows]
    failures = [r for r in rows if str(r.get("success", "")).lower() != "true"]

    min_ts = min(timestamps)
    max_ts = max(timestamps)
    total_seconds = max((max_ts - min_ts) / 1000.0, 1.0)
    total = len(rows)
    failed = len(failures)

    return {
        "total": total,
        "failed": failed,
        "error_pct": (failed / total) * 100 if total else 0.0,
        "avg_ms": sum(elapsed) / total if total else 0.0,
        "avg_latency_ms": sum(latency) / total if total else 0.0,
        "p90_ms": percentile(elapsed, 0.9),
        "rps": total / total_seconds if total_seconds else 0.0,
        "min_ts": min_ts,
        "max_ts": max_ts,
    }


def group_rows_by_label(rows):
    grouped = defaultdict(list)
    for row in rows:
        grouped[row.get("label", "unknown")] += [row]
    return grouped


def build_status(metrics):
    status = "passed"
    status_details = {"message": "Execucao concluida dentro do SLO esperado."}

    if metrics["failed"] > 0:
        status = "failed"
        status_details = {"message": f"Foram detectadas falhas funcionais: {metrics['failed']} amostras com success=false."}
    elif metrics["p90_ms"] >= ACCEPTANCE_P90_MS or metrics["rps"] < ACCEPTANCE_RPS:
        status = "failed"
        status_details = {
            "message": (
                f"SLO nao atendido. rps={metrics['rps']:.2f} (meta >= {ACCEPTANCE_RPS:.0f}), "
                f"p90={metrics['p90_ms']:.2f}ms (meta < {ACCEPTANCE_P90_MS}ms)."
            )
        }

    return status, status_details


def write_attachment(output_dir: Path, name: str, content: str, mime_type: str):
    suffix = ".txt"
    if mime_type == "application/json":
        suffix = ".json"
    source = f"{uuid.uuid4()}-attachment{suffix}"
    (output_dir / source).write_text(content, encoding="utf-8")
    return {"name": name, "source": source, "type": mime_type}


def write_allure_result(output_dir: Path, result: dict):
    result_file = output_dir / f"{uuid.uuid4()}-result.json"
    result_file.write_text(json.dumps(result, ensure_ascii=True, indent=2), encoding="utf-8")


def build_common_labels(test_name):
    return [
        {"name": "framework", "value": "jmeter"},
        {"name": "language", "value": "groovy"},
        {"name": "suite", "value": "BlazeDemo"},
        {"name": "parentSuite", "value": "Performance"},
        {"name": "epic", "value": "BlazeDemo"},
        {"name": "feature", "value": "Checkout"},
        {"name": "story", "value": test_name},
    ]


def write_test_summary_case(output_dir: Path, test_name: str, source_file: str, metrics: dict, rows):
    status, status_details = build_status(metrics)
    labels = build_common_labels(test_name)
    labels += [{"name": "severity", "value": "critical" if status == "failed" else "normal"}]

    failed_rows = [r for r in rows if str(r.get("success", "")).lower() != "true"][:TOP_FAILED_SAMPLES]
    sample_errors = [
        {
            "label": r.get("label", "unknown"),
            "responseCode": r.get("responseCode", ""),
            "responseMessage": r.get("responseMessage", ""),
            "elapsed": safe_int(r.get("elapsed", "0")),
            "url": r.get("URL", ""),
        }
        for r in failed_rows
    ]

    attachments = [
        write_attachment(
            output_dir,
            "metrics-summary",
            json.dumps(
                {
                    "test_name": test_name,
                    "source_file": source_file,
                    "target_url": TARGET_URL,
                    "total_samples": metrics["total"],
                    "failed_samples": metrics["failed"],
                    "error_pct": round(metrics["error_pct"], 4),
                    "avg_ms": round(metrics["avg_ms"], 2),
                    "avg_latency_ms": round(metrics["avg_latency_ms"], 2),
                    "p90_ms": round(metrics["p90_ms"], 2),
                    "rps": round(metrics["rps"], 2),
                    "acceptance_rps": ACCEPTANCE_RPS,
                    "acceptance_p90_ms": ACCEPTANCE_P90_MS,
                },
                ensure_ascii=True,
                indent=2,
            ),
            "application/json",
        ),
        write_attachment(
            output_dir,
            "failed-sample-extract",
            json.dumps(sample_errors, ensure_ascii=True, indent=2),
            "application/json",
        ),
    ]

    result = {
        "uuid": str(uuid.uuid4()),
        "name": test_name,
        "fullName": f"BlazeDemo::{test_name}",
        "historyId": test_name.replace(" ", "_").lower(),
        "status": status,
        "statusDetails": status_details,
        "stage": "finished",
        "start": metrics["min_ts"],
        "stop": metrics["max_ts"],
        "labels": labels,
        "parameters": [
            {"name": "target_url", "value": TARGET_URL},
            {"name": "source_file", "value": source_file},
            {"name": "total_samples", "value": str(metrics["total"])},
            {"name": "failed_samples", "value": str(metrics["failed"])},
            {"name": "error_pct", "value": f"{metrics['error_pct']:.4f}"},
            {"name": "avg_ms", "value": f"{metrics['avg_ms']:.2f}"},
            {"name": "avg_latency_ms", "value": f"{metrics['avg_latency_ms']:.2f}"},
            {"name": "p90_ms", "value": f"{metrics['p90_ms']:.2f}"},
            {"name": "rps", "value": f"{metrics['rps']:.2f}"},
            {"name": "acceptance_rps", "value": f"{ACCEPTANCE_RPS:.0f}"},
            {"name": "acceptance_p90_ms", "value": str(ACCEPTANCE_P90_MS)},
        ],
        "attachments": attachments,
    }
    write_allure_result(output_dir, result)


def write_label_cases(output_dir: Path, test_name: str, grouped_rows):
    for label, rows in grouped_rows.items():
        metrics = compute_metrics(rows)
        status, status_details = build_status(metrics)
        response_codes = Counter([r.get("responseCode", "unknown") for r in rows])
        errors_by_message = Counter(
            [r.get("responseMessage", "unknown") for r in rows if str(r.get("success", "")).lower() != "true"]
        )

        labels = build_common_labels(test_name)
        labels += [
            {"name": "subSuite", "value": "Samplers"},
            {"name": "story", "value": label},
            {"name": "severity", "value": "normal"},
        ]

        detail_payload = {
            "test_name": test_name,
            "sampler_label": label,
            "samples": metrics["total"],
            "failed": metrics["failed"],
            "error_pct": round(metrics["error_pct"], 4),
            "rps": round(metrics["rps"], 2),
            "avg_ms": round(metrics["avg_ms"], 2),
            "avg_latency_ms": round(metrics["avg_latency_ms"], 2),
            "p90_ms": round(metrics["p90_ms"], 2),
            "response_codes": response_codes,
            "top_error_messages": errors_by_message.most_common(5),
        }

        attachments = [
            write_attachment(output_dir, "sampler-metrics", json.dumps(detail_payload, ensure_ascii=True, indent=2), "application/json")
        ]

        result = {
            "uuid": str(uuid.uuid4()),
            "name": f"{test_name} :: {label}",
            "fullName": f"BlazeDemo::{test_name}::{label}",
            "historyId": f"{test_name}-{label}".replace(" ", "_").lower(),
            "status": status,
            "statusDetails": status_details,
            "stage": "finished",
            "start": metrics["min_ts"],
            "stop": metrics["max_ts"],
            "labels": labels,
            "parameters": [
                {"name": "sampler", "value": label},
                {"name": "samples", "value": str(metrics["total"])},
                {"name": "failed", "value": str(metrics["failed"])},
                {"name": "error_pct", "value": f"{metrics['error_pct']:.4f}"},
                {"name": "p90_ms", "value": f"{metrics['p90_ms']:.2f}"},
                {"name": "rps", "value": f"{metrics['rps']:.2f}"},
            ],
            "attachments": attachments,
        }
        write_allure_result(output_dir, result)


def write_environment_properties(output_dir: Path):
    env_file = output_dir / "environment.properties"
    existing = {}

    if env_file.exists():
        for line in env_file.read_text(encoding="utf-8").splitlines():
            if "=" in line:
                key, value = line.split("=", 1)
                existing[key.strip()] = value.strip()

    existing.update(
        {
            "target.url": TARGET_URL,
            "acceptance.rps": f"{ACCEPTANCE_RPS:.0f}",
            "acceptance.p90.ms": str(ACCEPTANCE_P90_MS),
            "jmeter.version": os.getenv("JMETER_VERSION", "5.6.3"),
            "python.version": sys.version.split()[0],
            "runner.os": os.getenv("RUNNER_OS", os.name),
            "runner.arch": os.getenv("RUNNER_ARCH", os.getenv("PROCESSOR_ARCHITECTURE", "unknown")),
            "github.repository": os.getenv("GITHUB_REPOSITORY", "local"),
            "github.ref": os.getenv("GITHUB_REF_NAME", "local"),
            "github.sha": os.getenv("GITHUB_SHA", "local"),
        }
    )

    lines = [f"{k}={v}" for k, v in sorted(existing.items())]
    env_file.write_text("\n".join(lines) + "\n", encoding="utf-8")


def write_executor_json(output_dir: Path):
    executor = {
        "name": "GitHub Actions" if os.getenv("GITHUB_ACTIONS") == "true" else "Local Execution",
        "type": "github",
        "url": os.getenv("GITHUB_SERVER_URL", ""),
        "buildOrder": safe_int(os.getenv("GITHUB_RUN_NUMBER", "0")),
        "buildName": os.getenv("GITHUB_WORKFLOW", "local-run"),
        "buildUrl": (
            f"{os.getenv('GITHUB_SERVER_URL', '')}/{os.getenv('GITHUB_REPOSITORY', '')}/actions/runs/"
            f"{os.getenv('GITHUB_RUN_ID', '')}"
            if os.getenv("GITHUB_ACTIONS") == "true"
            else ""
        ),
        "reportUrl": "",
        "reportName": "BlazeDemo performance",
    }
    (output_dir / "executor.json").write_text(json.dumps(executor, ensure_ascii=True, indent=2), encoding="utf-8")


def write_categories_json(output_dir: Path):
    categories = [
        {
            "name": "SLO breach (latency/throughput)",
            "matchedStatuses": ["failed"],
            "messageRegex": "SLO nao atendido.*",
        },
        {
            "name": "Functional failure",
            "matchedStatuses": ["failed"],
            "messageRegex": "Foram detectadas falhas funcionais.*",
        },
        {
            "name": "Infrastructure timeout/network",
            "matchedStatuses": ["broken", "failed"],
            "traceRegex": ".*(SocketTimeoutException|Read timed out|ConnectException).*",
        },
    ]
    (output_dir / "categories.json").write_text(json.dumps(categories, ensure_ascii=True, indent=2), encoding="utf-8")


def write_summary_file(output_dir: Path, test_name: str, metrics: dict):
    summary = {
        "test_name": test_name,
        "samples": metrics["total"],
        "failed": metrics["failed"],
        "error_pct": round(metrics["error_pct"], 4),
        "avg_ms": round(metrics["avg_ms"], 2),
        "avg_latency_ms": round(metrics["avg_latency_ms"], 2),
        "p90_ms": round(metrics["p90_ms"], 2),
        "rps": round(metrics["rps"], 2),
        "acceptance_rps": ACCEPTANCE_RPS,
        "acceptance_p90_ms": ACCEPTANCE_P90_MS,
    }
    (output_dir / f"{test_name.replace(' ', '_').lower()}-summary.json").write_text(
        json.dumps(summary, ensure_ascii=True, indent=2),
        encoding="utf-8",
    )
    print(json.dumps(summary, ensure_ascii=True))


def main():
    if len(sys.argv) != 4:
        print("Uso: python scripts/jtl_to_allure.py <jtl_input> <allure_results_dir> <test_name>")
        sys.exit(1)

    jtl_input = Path(sys.argv[1])
    allure_results = Path(sys.argv[2])
    test_name = sys.argv[3]

    if not jtl_input.exists():
        print(f"Nao foi possivel encontrar o JTL: {jtl_input}")
        sys.exit(2)

    allure_results.mkdir(parents=True, exist_ok=True)

    with jtl_input.open("r", encoding="utf-8", newline="") as fp:
        reader = csv.DictReader(fp)
        rows = list(reader)

    metrics = compute_metrics(rows)
    grouped = group_rows_by_label(rows)

    write_test_summary_case(allure_results, test_name, str(jtl_input), metrics, rows)
    write_label_cases(allure_results, test_name, grouped)
    write_environment_properties(allure_results)
    write_executor_json(allure_results)
    write_categories_json(allure_results)
    write_summary_file(allure_results, test_name, metrics)


if __name__ == "__main__":
    main()

File: jtl-allure/src/main/java/com/blazedemo/perf/JtlToAllure.java
package com.blazedemo.perf;

import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

public final class JtlToAllure {

    public static final int EXIT_ACCEPTANCE_FAILED = 3;

    private JtlToAllure() {}

    public static void run(Path jtlInput, Path allureResults, String testName, Config config) throws IOException {
        List<Map<String, String>> rows = JtlCsvReader.readAll(jtlInput);
        MetricsResult metrics = MetricsCalculator.computeMetrics(rows);
        Map<String, List<Map<String, String>>> grouped = groupRowsByLabel(rows);

        Files.createDirectories(allureResults);
        writeTestSummaryCase(allureResults, testName, jtlInput.toString(), metrics, rows, config);
        writeLabelCases(allureResults, testName, grouped, config);
        writeEnvironmentProperties(allureResults, config);
        writeExecutorJson(allureResults);
        writeCategoriesJson(allureResults);
        writeSummaryFile(allureResults, testName, metrics, config);

        StatusResult sr = buildStatus(metrics, config);
        if ("failed".equals(sr.status()) && config.strictAcceptance) {
            System.err.println(sr.message());
            System.exit(EXIT_ACCEPTANCE_FAILED);
        }
    }

    private record StatusResult(String status, String message) {}

    private static StatusResult buildStatus(MetricsResult m, Config config) {
        if (m.failed() > 0) {
            return new StatusResult(
                    "failed",
                    "Foram detectadas falhas funcionais: " + m.failed() + " amostras com success=false."
            );
        }
        if (m.p90Ms() >= config.acceptanceP90Ms || m.rps() < config.acceptanceRps) {
            return new StatusResult(
                    "failed",
                    String.format(
                            Locale.ROOT,
                            "SLO nao atendido. rps=%.2f (meta >= %.0f), p90=%.2fms (meta < %dms).",
                            m.rps(),
                            config.acceptanceRps,
                            m.p90Ms(),
                            config.acceptanceP90Ms
                    )
            );
        }
        return new StatusResult("passed", "Execucao concluida dentro do SLO esperado.");
    }

    private static Map<String, List<Map<String, String>>> groupRowsByLabel(List<Map<String, String>> rows) {
        Map<String, List<Map<String, String>>> grouped = new LinkedHashMap<>();
        for (Map<String, String> r : rows) {
            String label = r.getOrDefault("label", "unknown");
            grouped.computeIfAbsent(label, k -> new ArrayList<>()).add(r);
        }
        return grouped;
    }

    private static Map<String, Object> writeAttachment(Path outputDir, String name, String content, String mimeType)
            throws IOException {
        String suffix = "application/json".equals(mimeType) ? ".json" : ".txt";
        String source = UUID.randomUUID() + "-attachment" + suffix;
        Files.writeString(outputDir.resolve(source), content, StandardCharsets.UTF_8);
        Map<String, Object> att = new HashMap<>();
        att.put("name", name);
        att.put("source", source);
        att.put("type", mimeType);
        return att;
    }

    private static void writeAllureResult(Path outputDir, ObjectNode result) throws IOException {
        String file = UUID.randomUUID() + "-result.json";
        Files.writeString(outputDir.resolve(file), JsonHelper.mapper().writeValueAsString(result), StandardCharsets.UTF_8);
    }

    private static List<Map<String, String>> buildCommonLabels(String testName) {
        List<Map<String, String>> labels = new ArrayList<>();
        labels.add(Map.of("name", "framework", "value", "jmeter"));
        labels.add(Map.of("name", "language", "value", "groovy"));
        labels.add(Map.of("name", "suite", "value", "BlazeDemo"));
        labels.add(Map.of("name", "parentSuite", "value", "Performance"));
        labels.add(Map.of("name", "epic", "value", "BlazeDemo"));
        labels.add(Map.of("name", "feature", "value", "Checkout"));
        labels.add(Map.of("name", "story", "value", testName));
        return labels;
    }

    private static void writeTestSummaryCase(
            Path outputDir,
            String testName,
            String sourceFile,
            MetricsResult metrics,
            List<Map<String, String>> rows,
            Config config
    ) throws IOException {
        StatusResult sr = buildStatus(metrics, config);
        List<Map<String, String>> labels = new ArrayList<>(buildCommonLabels(testName));
        Map<String, String> sev = new HashMap<>();
        sev.put("name", "severity");
        sev.put("value", "failed".equals(sr.status()) ? "critical" : "normal");
        labels.add(sev);

        List<Map<String, String>> failedRows = rows.stream()
                .filter(r -> {
                    String s = r.get("success");
                    return s == null || !s.toLowerCase(Locale.ROOT).equals("true");
                })
                .limit(config.topFailedSamples)
                .collect(Collectors.toList());

        List<Map<String, Object>> sampleErrors = new ArrayList<>();
        for (Map<String, String> r : failedRows) {
            Map<String, Object> err = new HashMap<>();
            err.put("label", r.getOrDefault("label", "unknown"));
            err.put("responseCode", r.getOrDefault("responseCode", ""));
            err.put("responseMessage", r.getOrDefault("responseMessage", ""));
            err.put("elapsed", MetricsCalculator.safeInt(r.get("elapsed"), 0));
            err.put("url", r.getOrDefault("URL", ""));
            sampleErrors.add(err);
        }

        var metricsSummary = new HashMap<String, Object>();
        metricsSummary.put("test_name", testName);
        metricsSummary.put("source_file", sourceFile);
        metricsSummary.put("target_url", config.targetUrl);
        metricsSummary.put("total_samples", metrics.total());
        metricsSummary.put("failed_samples", metrics.failed());
        metricsSummary.put("error_pct", round4(metrics.errorPct()));
        metricsSummary.put("avg_ms", round2(metrics.avgMs()));
        metricsSummary.put("avg_latency_ms", round2(metrics.avgLatencyMs()));
        metricsSummary.put("p90_ms", round2(metrics.p90Ms()));
        metricsSummary.put("rps", round2(metrics.rps()));
        metricsSummary.put("acceptance_rps", config.acceptanceRps);
        metricsSummary.put("acceptance_p90_ms", config.acceptanceP90Ms);

        List<Map<String, Object>> attachments = new ArrayList<>();
        attachments.add(writeAttachment(
                outputDir,
                "metrics-summary",
                JsonHelper.mapper().writerWithDefaultPrettyPrinter().writeValueAsString(metricsSummary),
                "application/json"
        ));
        attachments.add(writeAttachment(
                outputDir,
                "failed-sample-extract",
                JsonHelper.mapper().writerWithDefaultPrettyPrinter().writeValueAsString(sampleErrors),
                "application/json"
        ));

        ObjectNode statusDetails = JsonHelper.mapper().createObjectNode();
        statusDetails.put("message", sr.message());

        ObjectNode result = JsonHelper.mapper().createObjectNode();
        result.put("uuid", UUID.randomUUID().toString());
        result.put("name", testName);
        result.put("fullName", "BlazeDemo::" + testName);
        result.put("historyId", testName.replace(" ", "_").toLowerCase(Locale.ROOT));
        result.put("status", sr.status());
        result.set("statusDetails", statusDetails);
        result.put("stage", "finished");
        result.put("start", metrics.minTs());
        result.put("stop", metrics.maxTs());
        result.set("labels", JsonHelper.mapper().valueToTree(labels));

        ArrayNode parameters = JsonHelper.mapper().createArrayNode();
        parameters.add(param("target_url", config.targetUrl));
        parameters.add(param("source_file", sourceFile));
        parameters.add(param("total_samples", String.valueOf(metrics.total())));
        parameters.add(param("failed_samples", String.valueOf(metrics.failed())));
        parameters.add(param("error_pct", String.format(Locale.ROOT, "%.4f", metrics.errorPct())));
        parameters.add(param("avg_ms", String.format(Locale.ROOT, "%.2f", metrics.avgMs())));
        parameters.add(param("avg_latency_ms", String.format(Locale.ROOT, "%.2f", metrics.avgLatencyMs())));
        parameters.add(param("p90_ms", String.format(Locale.ROOT, "%.2f", metrics.p90Ms())));
        parameters.add(param("rps", String.format(Locale.ROOT, "%.2f", metrics.rps())));
        parameters.add(param("acceptance_rps", String.format(Locale.ROOT, "%.0f", config.acceptanceRps)));
        parameters.add(param("acceptance_p90_ms", String.valueOf(config.acceptanceP90Ms)));
        result.set("parameters", parameters);
        result.set("attachments", JsonHelper.mapper().valueToTree(attachments));

        writeAllureResult(outputDir, result);
    }

    private static ObjectNode param(String name, String value) {
        ObjectNode o = JsonHelper.mapper().createObjectNode();
        o.put("name", name);
        o.put("value", value);
        return o;
    }

    private static void writeLabelCases(
            Path outputDir,
            String testName,
            Map<String, List<Map<String, String>>> grouped,
            Config config
    ) throws IOException {
        for (Map.Entry<String, List<Map<String, String>>> e : grouped.entrySet()) {
            String label = e.getKey();
            List<Map<String, String>> labelRows = e.getValue();
            MetricsResult metrics = MetricsCalculator.computeMetrics(labelRows);
            StatusResult sr = buildStatus(metrics, config);

            Map<String, Long> responseCodes = new HashMap<>();
            Map<String, Long> errorsByMessage = new HashMap<>();
            for (Map<String, String> r : labelRows) {
                String code = r.getOrDefault("responseCode", "unknown");
                responseCodes.merge(code, 1L, Long::sum);
                String success = r.get("success");
                if (success == null || !success.toLowerCase(Locale.ROOT).equals("true")) {
                    String msg = r.getOrDefault("responseMessage", "unknown");
                    errorsByMessage.merge(msg, 1L, Long::sum);
                }
            }

            List<List<Object>> topErrorMessages = errorsByMessage.entrySet().stream()
                    .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
                    .limit(5)
                    .map(en -> List.<Object>of(en.getKey(), en.getValue()))
                    .collect(Collectors.toList());

            List<Map<String, String>> labels = new ArrayList<>(buildCommonLabels(testName));
            labels.add(Map.of("name", "subSuite", "value", "Samplers"));
            labels.add(Map.of("name", "story", "value", label));
            labels.add(Map.of("name", "severity", "value", "normal"));

            Map<String, Object> detailPayload = new HashMap<>();
            detailPayload.put("test_name", testName);
            detailPayload.put("sampler_label", label);
            detailPayload.put("samples", metrics.total());
            detailPayload.put("failed", metrics.failed());
            detailPayload.put("error_pct", round4(metrics.errorPct()));
            detailPayload.put("rps", round2(metrics.rps()));
            detailPayload.put("avg_ms", round2(metrics.avgMs()));
            detailPayload.put("avg_latency_ms", round2(metrics.avgLatencyMs()));
            detailPayload.put("p90_ms", round2(metrics.p90Ms()));
            detailPayload.put("response_codes", responseCodes);
            detailPayload.put("top_error_messages", topErrorMessages);

            List<Map<String, Object>> attachments = new ArrayList<>();
            attachments.add(writeAttachment(
                    outputDir,
                    "sampler-metrics",
                    JsonHelper.mapper().writerWithDefaultPrettyPrinter().writeValueAsString(detailPayload),
                    "application/json"
            ));

            ObjectNode statusDetails = JsonHelper.mapper().createObjectNode();
            statusDetails.put("message", sr.message());

            ObjectNode result = JsonHelper.mapper().createObjectNode();
            result.put("uuid", UUID.randomUUID().toString());
            result.put("name", testName + " :: " + label);
            result.put("fullName", "BlazeDemo::" + testName + "::" + label);
            result.put("historyId", (testName + "-" + label).replace(" ", "_").toLowerCase(Locale.ROOT));
            result.put("status", sr.status());
            result.set("statusDetails", statusDetails);
            result.put("stage", "finished");
            result.put("start", metrics.minTs());
            result.put("stop", metrics.maxTs());
            result.set("labels", JsonHelper.mapper().valueToTree(labels));

            ArrayNode parameters = JsonHelper.mapper().createArrayNode();
            parameters.add(param("sampler", label));
            parameters.add(param("samples", String.valueOf(metrics.total())));
            parameters.add(param("failed", String.valueOf(metrics.failed())));
            parameters.add(param("error_pct", String.format(Locale.ROOT, "%.4f", metrics.errorPct())));
            parameters.add(param("p90_ms", String.format(Locale.ROOT, "%.2f", metrics.p90Ms())));
            parameters.add(param("rps", String.format(Locale.ROOT, "%.2f", metrics.rps())));
            result.set("parameters", parameters);
            result.set("attachments", JsonHelper.mapper().valueToTree(attachments));

            writeAllureResult(outputDir, result);
        }
    }

    private static void writeEnvironmentProperties(Path outputDir, Config config) throws IOException {
        Path envFile = outputDir.resolve("environment.properties");
        Map<String, String> existing = new HashMap<>();
        if (Files.exists(envFile)) {
            for (String line : Files.readAllLines(envFile, StandardCharsets.UTF_8)) {
                int eq = line.indexOf('=');
                if (eq > 0) {
                    existing.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
                }
            }
        }
        String newRps = String.format(Locale.ROOT, "%.0f", config.acceptanceRps);
        if (existing.containsKey("acceptance.rps") && !newRps.equals(existing.get("acceptance.rps"))) {
            existing.put("acceptance.rps.load", existing.remove("acceptance.rps"));
            existing.put("acceptance.rps.peak", newRps);
        } else {
            existing.put("acceptance.rps", newRps);
        }
        existing.put("target.url", config.targetUrl);
        existing.put("acceptance.p90.ms", String.valueOf(config.acceptanceP90Ms));
        existing.put("jmeter.version", envOrDefault("JMETER_VERSION", "5.6.3"));
        existing.put("java.version", System.getProperty("java.version", "unknown"));
        existing.put("runner.os", envOrDefault("RUNNER_OS", System.getProperty("os.name", "")));
        existing.put("runner.arch", envOrDefault("RUNNER_ARCH", envOrDefault("PROCESSOR_ARCHITECTURE", "unknown")));
        existing.put("github.repository", envOrDefault("GITHUB_REPOSITORY", "local"));
        existing.put("github.ref", envOrDefault("GITHUB_REF_NAME", "local"));
        existing.put("github.sha", envOrDefault("GITHUB_SHA", "local"));

        List<String> lines = existing.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .map(en -> en.getKey() + "=" + en.getValue())
                .collect(Collectors.toList());
        Files.writeString(envFile, String.join("\n", lines) + "\n", StandardCharsets.UTF_8);
    }

    private static String envOrDefault(String key, String def) {
        String v = System.getenv(key);
        return v == null || v.isEmpty() ? def : v;
    }

    private static void writeExecutorJson(Path outputDir) throws IOException {
        boolean gh = "true".equalsIgnoreCase(System.getenv("GITHUB_ACTIONS"));
        ObjectNode executor = JsonHelper.mapper().createObjectNode();
        executor.put("name", gh ? "GitHub Actions" : "Local Execution");
        executor.put("type", "github");
        executor.put("url", envOrDefault("GITHUB_SERVER_URL", ""));
        executor.put("buildOrder", MetricsCalculator.safeInt(System.getenv("GITHUB_RUN_NUMBER"), 0));
        executor.put("buildName", envOrDefault("GITHUB_WORKFLOW", "local-run"));
        String buildUrl = "";
        if (gh) {
            String server = envOrDefault("GITHUB_SERVER_URL", "");
            String repo = envOrDefault("GITHUB_REPOSITORY", "");
            String runId = envOrDefault("GITHUB_RUN_ID", "");
            buildUrl = server + "/" + repo + "/actions/runs/" + runId;
        }
        executor.put("buildUrl", buildUrl);
        executor.put("reportUrl", "");
        executor.put("reportName", "BlazeDemo performance");
        Files.writeString(
                outputDir.resolve("executor.json"),
                JsonHelper.mapper().writerWithDefaultPrettyPrinter().writeValueAsString(executor),
                StandardCharsets.UTF_8
        );
    }

    private static void writeCategoriesJson(Path outputDir) throws IOException {
        ArrayNode categories = JsonHelper.mapper().createArrayNode();
        ObjectNode c1 = JsonHelper.mapper().createObjectNode();
        c1.put("name", "SLO breach (latency/throughput)");
        c1.set("matchedStatuses", JsonHelper.mapper().valueToTree(List.of("failed")));
        c1.put("messageRegex", "SLO nao atendido.*");
        categories.add(c1);
        ObjectNode c2 = JsonHelper.mapper().createObjectNode();
        c2.put("name", "Functional failure");
        c2.set("matchedStatuses", JsonHelper.mapper().valueToTree(List.of("failed")));
        c2.put("messageRegex", "Foram detectadas falhas funcionais.*");
        categories.add(c2);
        ObjectNode c3 = JsonHelper.mapper().createObjectNode();
        c3.put("name", "Infrastructure timeout/network");
        c3.set("matchedStatuses", JsonHelper.mapper().valueToTree(List.of("broken", "failed")));
        c3.put("traceRegex", ".*(SocketTimeoutException|Read timed out|ConnectException).*");
        categories.add(c3);
        Files.writeString(
                outputDir.resolve("categories.json"),
                JsonHelper.mapper().writerWithDefaultPrettyPrinter().writeValueAsString(categories),
                StandardCharsets.UTF_8
        );
    }

    private static void writeSummaryFile(Path outputDir, String testName, MetricsResult metrics, Config config)
            throws IOException {
        ObjectNode summary = JsonHelper.mapper().createObjectNode();
        summary.put("test_name", testName);
        summary.put("samples", metrics.total());
        summary.put("failed", metrics.failed());
        summary.put("error_pct", round4(metrics.errorPct()));
        summary.put("avg_ms", round2(metrics.avgMs()));
        summary.put("avg_latency_ms", round2(metrics.avgLatencyMs()));
        summary.put("p90_ms", round2(metrics.p90Ms()));
        summary.put("rps", round2(metrics.rps()));
        summary.put("acceptance_rps", config.acceptanceRps);
        summary.put("acceptance_p90_ms", config.acceptanceP90Ms);

        String filename = testName.replace(" ", "_").toLowerCase(Locale.ROOT) + "-summary.json";
        var compact = JsonHelper.mapper().copy().disable(SerializationFeature.INDENT_OUTPUT);
        String json = compact.writeValueAsString(summary);
        Files.writeString(outputDir.resolve(filename), json, StandardCharsets.UTF_8);
        System.out.println(json);
    }

    private static double round2(double v) {
        return Math.round(v * 100.0) / 100.0;
    }

    private static double round4(double v) {
        return Math.round(v * 10000.0) / 10000.0;
    }
}
